### PR TITLE
Reconfigure ZHA device take 2

### PR DESCRIPTION
**Pull request by dmulcahey**
## Proposed change

This PR enhances the new reconfigure device dialog for ZHA devices. The log relaying has been removed and it has been replaced by a more user friendly dialog.

https://user-images.githubusercontent.com/1335687/116247627-0c85c580-a739-11eb-839e-ae8e22901ff5.mov

**REQUIRES** https://github.com/home-assistant/core/pull/49672

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by dmulcahey** [comment]
@bramkragten right now I am kind of hacking _eventCount because when map properties change they don't appear to actually cause re-rendering. I am assuming there is a better way to do this?

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts [hunk 1]
```diff
@@ -8,42 +8,75 @@ import {
...
+      )
+    );
     this._subscribe(params);
   }
 
   public closeDialog(): void {
     this._unsubscribe();
-    this._formattedEvents = "";
     this._params = undefined;
+    this._clusterConfigurationStatuses = new Map();
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts:78]:
```suggestion
    this._clusterConfigurationStatuses = undefined;
```

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts [hunk 2]
```diff
@@ -56,53 +89,194 @@ class DialogZHAReconfigureDevice extends LitElement {
...
-  private _handleMessage(message: any): void {
-    if (message.type === LOG_OUTPUT) {
-      this._formattedEvents += message.log_entry.message + "\n";
-      const paperTextArea = this.shadowRoot!.querySelector("paper-textarea");
-      if (paperTextArea) {
-        const textArea = (paperTextArea.inputElement as IronAutogrowTextareaElement)
-          .textarea;
-        textArea.scrollTop = textArea.scrollHeight;
+  private _handleMessage(message: ClusterConfigurationEvent): void {
+    // this is currently here to hack rerendering because map updates aren't triggering rendering?
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts:254]: You will need to make a shallow copy of the map to make it trigger rendering:
`this._clusterConfigurationStatuses = new Map(this._clusterConfigurationStatuses)`
Or call `this.requestUpdate()`

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts [hunk 3]
```diff
@@ -8,101 +8,385 @@ import {
...
+  @internalProperty() private _allSuccessful = true;
 
-  private _reconfigureDeviceTimeoutHandle: any = undefined;
+  @internalProperty() private _showDetails = false;
+
+  private _subscribed?: Promise<() => Promise<void>>;
 
   public async showDialog(
     params: ZHAReconfigureDeviceDialogParams
   ): Promise<void> {
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts:58]:
```suggestion
  public showDialog(
    params: ZHAReconfigureDeviceDialogParams
  ): void {
```

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts [hunk 4]
```diff
@@ -8,101 +8,385 @@ import {
...
+          </span>
+          <mwc-icon-button
+            aria-label=${this.hass.localize("ui.dialogs.generic.close")}
+            dialogAction="close"
+            class="header_button"
+            dir=${computeRTLDirection(this.hass)}
+          >
+            <ha-svg-icon .path=${mdiClose}></ha-svg-icon>
+          </mwc-icon-button>
+        `}
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts:94]: Why not use `createCloseHeading`?

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts [hunk 5]
```diff
@@ -8,101 +8,385 @@ import {
...
+          },
+        ]
+      )
+    );
+    this._subscribe(this._params);
+    this._status = "started";
+  }
+
+  private _handleMessage(message: ClusterConfigurationEvent): void {
+    // this is currently here to hack rerendering because map updates aren't triggering rendering?
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts:358]:
```suggestion
```

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts [hunk 6]
```diff
@@ -8,101 +8,385 @@ import {
...
+  @internalProperty() private _params:
+    | ZHAReconfigureDeviceDialogParams
+    | undefined = undefined;
+
+  @internalProperty() private _allSuccessful = true;
 
-  private _reconfigureDeviceTimeoutHandle: any = undefined;
+  @internalProperty() private _showDetails = false;
+
+  private _subscribed?: Promise<() => Promise<void>>;
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts:54]:
```suggestion
  private _subscribed?: Promise<UnsubscribeFunc>;
```

**Review by bramkragten — APPROVED** [review]

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts (hunk 4 above)
Comment (dmulcahey) [reply src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts:94]: I think I had something else in there at some point... I had the spinner in it that's what it was. I can fix this

Comment (dmulcahey) [reply src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts]: done

File: src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts [hunk 7]
```diff
@@ -8,101 +8,374 @@ import {
...
         open
-        hideActions
-        @closing="${this.closeDialog}"
+        @closed="${this.closeDialog}"
         .heading=${createCloseHeading(
           this.hass,
-          this.hass.localize(`ui.dialogs.zha_reconfigure_device.heading`)
+          this.hass.localize(`ui.dialogs.zha_reconfigure_device.heading`) +
+            ": " +
+            (this._params?.device.user_given_name || this._params?.device.name)
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/zha/dialog-zha-reconfigure-device.ts:83]: `${this.hass.localize(`ui.dialogs.zha_reconfigure_device.heading`)}: ${(this._params?.device.user_given_name || this._params?.device.name)}`

**Pull request merged**